src/blr/rules/trim.smk

```diff
@@ -1,48 +1,70 @@
 """
 Rules connected to trimming of FASTQ files.
+
+READ1 LAYOUT
+
+5'-CAGTTGATCATCAGCAGGTAATCTGGBDVHBDVHBDVHBDVHBDVHCATGACCTCTTGGAACTGTCAGATGTGTATAAGAGACAGNNNN...NNNN(CTGTCTCTTATACACATCT)-3'
+   <------------h1----------><-------DBS--------><-----------------h2------------------><---gDNA--><---------h3-------->
+    h1 may inlude frameshift                                                                        Presence depencd on
+    oligos varying from 0-4                                                                         insert length
+    extra oligos in 5' end
+
+READ 2 LAYOUT
+
+5'-NNNN...NNNN(CTGTCTCTTATACACATCT)-3'
+   <---gDNA--><---------h3-------->
+              Presence depencd on
+              insert length
 """
 
+h1="CAGTTGATCATCAGCAGGTAATCTGG"
+DBS="N"*20
+h2="CATGACCTCTTGGAACTGTCAGATGTGTATAAGAGACAG"
+h3="CTGTCTCTTATACACATCT"
+trim_len=sum(map(len, [h1,DBS,h2]))
+extract_len=len(h1)
+
 rule trim:
-    # Trim away E handle on R1 5'. Also removes reads shorter than 85 bp.
-    # Extract barcode sequence and place in header.
-    # Cut H1691' + TES sequence from 5' of R1. H1691'=CATGACCTCTTGGAACTGTC, TES=AGATGTGTATAAGAGACAG.
-    # Cut 3' TES' sequence from R1 and R2. TES'=CTGTCTCTTATACACATCT
-    # Discard untrimmed.
+    # Trim away 5' and possible 3' handles on read1 and trim possible 3' handles on read2.
     output:
-        r1_fastq="trimmed-c.1.fastq.gz",
-        r2_fastq="trimmed-c.2.fastq.gz"
+        r1_fastq="trimmed.1.fastq.gz",
+        r2_fastq="trimmed.2.fastq.gz"
     input:
         r1_fastq="reads.1.fastq.gz",
-        r2_fastq="reads.2.fastq.gz"
-    log:
-        a = "trimmed-a.log",
-        b = "trimmed-b.log"
+        r2_fastq="reads.2.fastq.gz",
+    log: "cutadapt_trim.log"
     threads: 20
     shell:
-        "cutadapt"
-        " -g 'XNNNCAGTTGATCATCAGCAGGTAATCTGG;min_overlap=26'"
+        "cutadapt" #Initial trim
+        " -g 'XNNN{h1}{DBS}{h2};min_overlap={trim_len}...{h3};optional'"
+        " -A {h3}"
+        " --pair-filter 'first'"
         " -e 0.2"
         " --discard-untrimmed"
         " -j {threads}"
-        " -m 65"
-        " -o -"
-        " --interleaved"
+        " -m 25"
+        " -o {output.r1_fastq}"
+        " -p {output.r2_fastq}"
         " {input.r1_fastq}"
         " {input.r2_fastq}"
-        " 2> {log.a} |"
-        "blr extractbarcode"
-        " -"
-        " -o1 - |" 
-        "cutadapt"
-        " -a ^CATGACCTCTTGGAACTGTCAGATGTGTATAAGAGACAG...CTGTCTCTTATACACATCT"
-        " -A CTGTCTCTTATACACATCT"
+        "> {log} "
+
+rule extract_DBS:
+    # Extract barcode sequence from read1 FASTQ
+    output:
+        fastq="barcodes.fastq",
```

Comment:
Are the quality values actually needed here? If not, you could save some space (and I/O) by writing a `barcodes.fasta` file instead. Just change the name of the output file and Cutadapt will convert the format.

Reply:
From the starcode paper "raw sequence with count, FASTA or FASTQ (in which case starcode ignores the quality).", so fasta should be sufficient.

Reply:
I will change